### ISSUE TITLE
`MockMonitor.readResource` crashes when calling non-existent `getCustom` method on `resproto.ReadResourceRequest`

### ISSUE THREAD

**Issue by henriiik** [pulumi/pulumi#9111]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

When writing and running unit tests with pulumi using mocks, all calls to `MockMonitor.readResource` fail.

### Steps to reproduce

A small reproduction can be found here: https://github.com/henriiik/missing-get-custom

All that is required is to write a test that uses `pulumi.runtime.setMocks` and some `Resource.get` method, and run it. (my example uses the `@pulumi/aws` package because `@pulumi/pulumi` does not have any resource that has a `.get` that i know of)

```ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

pulumi.runtime.setMocks({
    newResource: function(args: pulumi.runtime.MockResourceArgs): {id: string, state: any} {
        return {
            id: args.inputs.name + "_id",
            state: args.inputs,
        };
    },
    call: function(args: pulumi.runtime.MockCallArgs) {
        return args.inputs;
    },
});

describe("aws.ec2.Vpc", function() {
    it("get", function(done) {
        let vpc = aws.ec2.Vpc.get("vpc-name","vpc-id");
        vpc.id.apply(id => {
            if (id !== "vpc-id") {
                done(`invalid vpc id! ${id}`);
            } else {
                done();
            }
        })
    })
});
```

Expected: the resource id should resolve to a value.
Actual: the test times out.

### Further details

When running the test with mocha, a promise created in the background is rejected and silently ignored, leading to the test timing out while waiting for an output that will never resolve.

When running the test with jest however, the test times out, but the unhandled rejected promise is logged to the console with an error message like this:

```
    failed to read resource #vpc-id 'vpc-name' [aws:ec2/vpc:Vpc]: req.getCustom is not a function
      at Object.readResource (node_modules/@pulumi/runtime/resource.ts:206:27)
      at new Resource (node_modules/@pulumi/resource.ts:387:13)
      at new CustomResource (node_modules/@pulumi/resource.ts:772:9)
      at new Vpc (node_modules/@pulumi/ec2/vpc.ts:210:9)
      at Function.get (node_modules/@pulumi/ec2/vpc.ts:57:16)
      at Object.<anonymous> (test.ts:24:27)
```

`MockMonitor.readResource` calls `req.getCustom`, here:

https://github.com/pulumi/pulumi/blob/a5ea2ef1e2a588fc6abc8f13ec06678b83233c48/sdk/nodejs/runtime/mocks.ts#L175

The type for `req` in `MockMonitor.readResource` is `any`, but in reality `req` is a `resproto.ReadResourceRequest`, which does not have a `getCustom` method. It is constructed here:

https://github.com/pulumi/pulumi/blob/a5ea2ef1e2a588fc6abc8f13ec06678b83233c48/sdk/nodejs/runtime/resource.ts#L213

And then then passed to `MockMonitor.readResource` here:

https://github.com/pulumi/pulumi/blob/a5ea2ef1e2a588fc6abc8f13ec06678b83233c48/sdk/nodejs/runtime/resource.ts#L236

**Comment by henriiik** [comment]
I created a work around this by subclassing MockMonitor:

```ts
import * as pulumi from "@pulumi/pulumi";
import * as mocks from "@pulumi/pulumi/runtime/mocks";

export class MockMonitor extends mocks.MockMonitor {
  readResource(req: any, callback: (err: any, innerResponse: any) => void): Promise<void> {
    if (req.getCustom === undefined) {
      // this is a workaround for https://github.com/pulumi/pulumi/issues/9111
      req.getCustom = () => undefined;
    }
    return super.readResource(req, callback);
  }
}

pulumi.runtime.setMockOptions(
  new MockMonitor({
    newResource: (args) => {
      // ...
    },
    call: (args) => {
      // ...
    },
  })
);
```

**Comment by justinvp** [comment]
Looks like this was fixed by #12072
